### PR TITLE
Http gateway message support

### PR DESCRIPTION
**Pull request by mfleisz**
This PR implements parsing of the consent and service messages in the http RD gateway implementation. The `PresentGatewayMessage` callback has only been called by the legacy gateway code.
Additionally the PR adds callback handlers for the Windows, Wayland and X11 clients.

### CODE REVIEW

File: client/Windows/wf_client.c [hunk 1]
```diff
@@ -638,6 +638,33 @@ static DWORD wf_verify_changed_certificate_ex(freerdp* instance, const char* hos
 	}
 }
 
+BOOL wf_present_gateway_message(freerdp* instance, UINT32 type, BOOL isDisplayMandatory,
```
Comment (akallabeth) [inline client/Windows/wf_client.c:641]: maybe `static`?

File: libfreerdp/core/gateway/rdg.c [hunk 2]
```diff
@@ -669,6 +711,76 @@ static BOOL rdg_process_handshake_response(rdpRdg* rdg, wStream* s)
...
+	{
+		WCHAR* msg;
+		UINT16 msgLenBytes;
+		rdpContext* context = rdg->context;
+
+		assert(context);
+		assert(context->instance);
+
+		/* Read message string and invoke callback */
+		if (!rdg_read_http_unicode_string(s, &msg, &msgLenBytes))
```
Comment (akallabeth) [inline libfreerdp/core/gateway/rdg.c:771]: Is this string really `\0` terminated? (unchecked data from server)

File: client/Windows/wf_client.c (hunk 1 above)
Comment (mfleisz) [reply client/Windows/wf_client.c]: Good point

File: libfreerdp/core/gateway/rdg.c (hunk 2 above)
Comment (mfleisz) [reply libfreerdp/core/gateway/rdg.c:771]: The documentation does not say anything about \0 termination but I check if there are at least msgLenBytes of string data available and the both parameters (buffer and buffer size) are forwarded to the callback

File: client/Windows/wf_client.c [hunk 3]
```diff
@@ -638,6 +638,33 @@ static DWORD wf_verify_changed_certificate_ex(freerdp* instance, const char* hos
 	}
 }
 
+static BOOL wf_present_gateway_message(freerdp* instance, UINT32 type, BOOL isDisplayMandatory,
```
Comment (akallabeth) [inline client/Windows/wf_client.c:641]: Just noticed, is there a documentation of what values `type` can have? An enum here would greatly improve readability (remove the magic numbers)

Comment (mfleisz) [reply client/Windows/wf_client.c:641]: @akallabeth Yes - they are actually already there (in core/gateway/tsg.h) so I will just export them to the public API. 👍

Comment (akallabeth) [reply client/Windows/wf_client.c:641]: @mfleisz great, then please also change the type for the callback accordingly.

Comment (akallabeth) [inline client/Windows/wf_client.c:641]: Wrong function signature, update to enum.

File: client/common/client.c [hunk 4]
```diff
@@ -663,6 +663,66 @@ DWORD client_cli_verify_changed_certificate_ex(freerdp* instance, const char* ho
 	return client_cli_accept_certificate(instance->settings);
 }
 
+BOOL client_cli_present_gateway_message(freerdp* instance, UINT32 type, BOOL isDisplayMandatory,
```
Comment (akallabeth) [inline client/common/client.c:666]: Wrong function signature, update to enum.

**Pull request merged**